File: solutions/csharp/protein-translation/1/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ProteinTranslation
{
    public static string[] Proteins(string strand)
    {
        return StrandIter(strand)
            .TakeWhile(x => x != null)
            .ToArray();
    }

    private static IEnumerable<string> StrandIter(string strand)
    {
        for (int i = 0; i < strand.Length; i += 3)
        {
            yield return strand.Substring(i, 3) switch
            {
                "AUG" => "Methionine",
                "UUU" or "UUC" => "Phenylalanine",
                "UUA" or "UUG" => "Leucine",
                "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
                "UAU" or "UAC"                   => "Tyrosine",
                "UGU" or "UGC"                   => "Cysteine",
                "UGG"                            => "Tryptophan",
                "UAA" or "UAG" or "UGA"          => null,
                _ => throw new InvalidOperationException(),
            };
        }
    }
}

File: solutions/typescript/protein-translation/1/protein-translation.ts
export function translate(strand: string) {
    return [...split_strand(strand)];
}

function* split_strand(strand: string) {
    for (let s = 0; s < strand.length; s += 3)
    {
        let rna = strand.slice(s, s + 3);

        switch (rna) {
            case "AUG":
                yield "Methionine"
                break;
            case "UUU":
            case "UUC":
                yield "Phenylalanine";
                break;
            case "UUA":
            case "UUG":
                yield "Leucine";
                break;
            case "UCU":
            case "UCC":
            case "UCA":
            case "UCG":
                yield "Serine";
                break;
            case "UAU":
            case "UAC":
                yield "Tyrosine";
                break;
            case "UGU":
            case "UGC":
                yield "Cysteine";
                break;
            case "UGG":
                yield "Tryptophan";
                break;
            case "UAA":
            case "UAG":
            case "UGA":
                return;
            default:
                throw new Error("Invalid codon");
        }
    }
}